primitives/src/merkle_tree/field_based_mht/smt/big_lazy_merkle_tree.rs

```diff
@@ -0,0 +1,1200 @@
+use algebra::{ToBytes, to_bytes, FromBytes};
+
+use crate::{
+    crh::{FieldBasedHash, BatchFieldBasedHash, FieldBasedHashParameters},
+    merkle_tree::{
+        field_based_mht::{
+            BatchFieldBasedMerkleTreeParameters, check_precomputed_parameters,
+            FieldBasedMerkleTreePath, FieldBasedBinaryMHTPath,
+            smt::{
+                Coord, OperationLeaf, ActionLeaf::Remove, BigMerkleTreeState, Error
+            },
+        },
+    }
+};
+
+use rocksdb::{DB, Options};
+
+use std::{
+  collections::HashSet, marker::PhantomData, fs, path::Path,
+};
+
+#[derive(Debug)]
+pub struct LazyBigMerkleTree<T: BatchFieldBasedMerkleTreeParameters> {
+    // if unset, all DBs and tree internal state will be deleted when an instance of this struct
+    // gets dropped
+    persistent: bool,
+    // path to state required to restore the tree
+    state_path: Option<String>,
+    // tree in-memory state
+    state: BigMerkleTreeState<T>,
+    // the number of leaves
+    width: usize,
+    // the height of the tree
+    height: usize,
+    // path to the db
+    path_db: String,
+    // stores the leaves
+    database: DB,
+    // path to the cache
+    path_cache: String,
+    // stores the cached nodes
+    db_cache: DB,
+
+    _parameters: PhantomData<T>,
+}
+
+impl<T: BatchFieldBasedMerkleTreeParameters> Drop for LazyBigMerkleTree<T> {
```

Comment:
Is this what we want ?

Reply:
I implemented the Drop method so that the DB is deleted. Otherwise, data will be persistent between different instances and affect the results between calls. There is another way that destroys the DB which is DestroyDB() but here I'm just deleting the directory.

Reply:
Yes, I'm referring to the new implementation here: so each time the instance corresponding to this struct goes out of scope, the DB and all the data will be destroyed/saved, according to the persistent flag value. But should we really implement a distructor like this ? Of course it depends on the use case, but probably would be best for the user to decide when closing everything. That's what I want to discuss about.

Reply:
Yes, this could be a good idea.